paasta_tools/kubernetes/bin/paasta_secrets_sync.py

```diff
@@ -245,63 +283,95 @@ def sync_secrets(
-                                )
-                            else:
-                                raise
-                        create_kubernetes_secret_signature(
-                            kube_client=kube_client,
-                            secret=secret,
-                            service=service,
-                            secret_signature=secret_signature,
-                            namespace=namespace,
-                        )
-                    elif secret_signature != kubernetes_secret_signature:
-                        log.info(
-                            f"{secret} for {service} in {namespace} needs updating as signature changed"
-                        )
-                        update_secret(
-                            kube_client=kube_client,
-                            secret=secret,
-                            service=service,
-                            secret_provider=secret_provider,
-                            namespace=namespace,
-                        )
-                        update_kubernetes_secret_signature(
-                            kube_client=kube_client,
-                            secret=secret,
-                            service=service,
-                            secret_signature=secret_signature,
-                            namespace=namespace,
-                        )
-                    else:
-                        log.info(f"{secret} for {service} in {namespace} up to date")
+                update_k8s_secret(
+                    service=service,
+                    signature_name=get_paasta_secret_signature_name(
+                        namespace, service, sanitise_kubernetes_name(secret)
+                    ),
+                    secret_name=get_paasta_secret_name(
+                        namespace, service, sanitise_kubernetes_name(secret)
+                    ),
+                    secret_data={
+                        secret: base64.b64encode(
+                            secret_provider.decrypt_secret_raw(secret)
+                        ).decode("utf-8")
+                    },
+                    secret_signature=secret_signature,
+                    kube_client=kube_client,
+                    namespace=namespace,
+                )
+
+    return True
+
+
+def sync_crypto_secrets(
+    kube_client: KubeClient,
+    cluster: str,
+    service: str,
+    secret_provider_name: str,
+    vault_cluster_config: Dict[str, str],
+    soa_dir: str,
+    vault_token_file: str,
+) -> bool:
+    """
+    For each key-name in `crypto_key`,
+    1. Fetch all versions of the key-name from Vault superregion mapped from cluster, e.g. `kubestage` maps to `devc` Vault server.
+    2. Create K8s secret from JSON blob containing all key versions.
+    3. Create signatures as K8s configmap based on JSON blob hash.
+
+    So each instance of a service get the same key, thereby reducing requests to Vault API
```

Comment:
is this still true? every instance has its own crypto_keys dict and we're calling `get_secret_provider(...).get_key_versions(...)` for every instance

Reply:
Worded it wrong: we are reducing API count for each instance/replica count of a service instance. We are fine making calls for each service instance, but we can probably cache results if soa-configs are same for instances.